File: src/main/java/no/nav/bidrag/commons/web/HttpResponse.java
package no.nav.bidrag.commons.web;

import java.util.Optional;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

/**
 * Redirecting {@link ResponseEntity} from one service to another
 *
 * @param <T> type of http payload
 */
public class HttpResponse<T> {

  private final ResponseEntity<T> responseEntity;

  public HttpResponse(ResponseEntity<T> responseEntity) {
    this.responseEntity = responseEntity;
  }

  public Optional<T> fetchBody() {
    return Optional.ofNullable(responseEntity.getBody());
  }

  public boolean is2xxSuccessful() {
    return responseEntity.getStatusCode().is2xxSuccessful();
  }

  public ResponseEntity<T> getResponseEntity() {
    return responseEntity;
  }

  public static <E> HttpResponse<E> from(HttpStatus httpStatus) {
    var responseEntity = new ResponseEntity<E>(httpStatus);
    return new HttpResponse<>(responseEntity);
  }

  public static <E> HttpResponse<E> from(HttpStatus httpStatus, E body) {
    var responseEntity = new ResponseEntity<>(body, httpStatus);
    return new HttpResponse<>(responseEntity);
  }
}

File: src/main/kotlin/no/nav/bidrag/commons/web/HttpResponse.kt
package no.nav.bidrag.commons.web

import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.util.CollectionUtils
import java.util.*

/**
 * Redirecting [ResponseEntity] from one service to another
 *
 * @param <T> type of http payload
</T> */
class HttpResponse<T>(val responseEntity: ResponseEntity<T>) {

    fun fetchBody(): Optional<T> {
        return Optional.ofNullable(responseEntity.body)
    }

    fun is2xxSuccessful(): Boolean {
        return responseEntity.statusCode.is2xxSuccessful
    }

    fun fetchHeaders(): HttpHeaders {
        return responseEntity.headers
    }

    fun clearContentHeaders(): HttpResponse<T?> {
        val headersMap: Map<String, List<String>> =
            responseEntity.headers
                .entries
                .associateBy({ (key) -> key }, { (_, value) -> value })

        val headers = HttpHeaders(CollectionUtils.toMultiValueMap(headersMap))
        headers.clearContentHeaders()
        return from(responseEntity.body, headers, responseEntity.statusCode)
    }

    companion object {
        fun <E> from(httpStatus: HttpStatusCode): HttpResponse<E> {
            val responseEntity = ResponseEntity<E>(httpStatus)
            return HttpResponse(responseEntity)
        }

        fun <E> from(httpStatus: HttpStatusCode, body: E): HttpResponse<E> {
            val responseEntity = ResponseEntity(body, httpStatus)
            return HttpResponse(responseEntity)
        }

        fun <E> from(httpHeaders: HttpHeaders, httpStatus: HttpStatusCode): HttpResponse<E> {
            val responseEntity = ResponseEntity<E>(httpHeaders, httpStatus)
            return HttpResponse(responseEntity)
        }

        fun <E> from(body: E, httpHeaders: HttpHeaders?, httpStatus: HttpStatusCode): HttpResponse<E> {
            val responseEntity = ResponseEntity(body, httpHeaders, httpStatus)
            return HttpResponse(responseEntity)
        }
    }
}
